src/lib.rs

```diff
@@ -72,45 +74,32 @@ impl State {
         let retry_enabled = self.config.retry_request_config;
         let retry_attempts = self.config.user_creation_retry_attempts;
 
-        let mut handles = FuturesUnordered::new();
-
         let progress_bar = create_progress_bar(
             "Init users".to_string(),
             (desired_users - actual_users).try_into().unwrap(),
         );
         progress_bar.tick();
 
-        for i in actual_users..desired_users {
-            handles.push(tokio::spawn({
-                let server = server.clone();
-                let metrics = self.metrics.clone();
-                let progress_bar = progress_bar.clone();
+        let range = (actual_users..desired_users).collect::<Vec<usize>>();
+
+        let futures = range.iter().map(|i| {
+            create_user(
+                server.clone(),
+                &progress_bar,
+                self.metrics.clone(),
+                *i,
+                retry_attempts,
+                timestamp,
+                retry_enabled,
+            )
+        });
 
-                async move {
-                    let id = format!("user_{i}_{timestamp}");
-                    for _ in 0..retry_attempts {
-                        let user = User::new(&id, &server, retry_enabled, metrics.clone()).await;
-
-                        if let Some(mut user) = user {
-                            if let Some(mut user) = user.register().await {
-                                if let Some(user) = user.login().await {
-                                    log::info!("User is now synching: {}", user.id());
-                                    progress_bar.inc(1);
-                                    return Some(user.sync().await);
-                                }
-                            }
-                        }
-                    }
-                    log::info!("Couldn't init a user");
-                    progress_bar.inc(1);
-                    None
-                }
-            }));
-        }
+        let stream_iter = futures::stream::iter(futures);
+        let mut buffered_iter = stream_iter.buffer_unordered(self.config.user_creation_throughput);
 
-        while let Some(user) = handles.next().await {
-            if let Ok(Some(user)) = user {
-                self.users.push(user)
+        while let Some(user) = buffered_iter.next().await {
+            if let Some(user) = user {
```

Comment:
If we are not able to create the number of users required, we should fail (as we have exhausted the retry attempts).

Reply:
Yes, we should panick in this case, it's a good task for the user creation issue I'd say OOS of this PR

Reply:
Ok, I'd say let's leave a `TODO` here to remind ourselves about it then.